### PR TITLE
switch to uuid-like revisions for firestore

### PR DESCRIPTION
**Pull request by fspmarshall**
This PR switches the firestore backend to use uuid-like revisions instead of firestore's builtin `UpdateTime`.  This switch is a prerequisite for implementing the `AtomicWrite` API for firestore.  Because firestore doesn't support non-interactive preconditions for values other than `UpdateTime`, this switch necessitates migrating the conditional APIs to use Firestore's transaction API.

The switch to the new revision style is done lazily rather than via migration.  All APIs continue to support the old revision, but all future writes add the new revision.  This change is safe for concurrent operation with older backend impls because adding the new revision field does not prevent continuing use of `UpdateTime`.

In order to verify the backwards compatibility of this feature, I created a fork of this branch that randomly applied revision v2 on only 50% of writes and ran through the test plan multiple times.  All tests passed seamlessly using a mix of revision styles.

### CODE REVIEW

**Review by rosstimothy — APPROVED** [review]

File: lib/backend/firestore/firestorebk.go [hunk 1]
```diff
@@ -549,41 +575,49 @@ func (b *Backend) CompareAndSwap(ctx context.Context, expected backend.Item, rep
...
-			Get(ctx)
+	replaceWith.Revision = createRevisionV2()
+	replaceRec := newRecord(replaceWith, b.clock)
+
+	docRef := b.svc.Collection(b.CollectionName).Doc(b.keyToDocumentID(replaceWith.Key))
+
+	err := b.svc.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
+		docSnap, err := tx.Get(docRef)
 		if err != nil {
 			err = ConvertGRPCError(err)
```
Comment (espadolini) [inline lib/backend/firestore/firestorebk.go:586]: Should we return unexpected firestore errors verbatim to RunTransaction? We seem to check for status code not found in ConditionalUpdate and ConditionalDelete.

Comment (fspmarshall) [reply lib/backend/firestore/firestorebk.go:586]: Performing error conversion inside of `RunTransaction` is safe (the firestore client treats all errors returned by the callback as opaque), and I think its more readable to perform error handling at the location we first observe the error.  The only code we check for outside of `RunTransaction` is the `Aborted` code, which is returned by `RunTrasnaction` itself, not by one of the specific methods within the callback.

**Review by Tener — APPROVED** [review]
Should we add some timelines for deprecating the legacy revision format?

**Review by espadolini — APPROVED** [review]
If we need to keep the old conditional code around anyway, why not treat the new revision mode as optional just for the atomic writes, and return the old one from reads and point writes?

**Comment by fspmarshall** [comment]
@espadolini I toy a bit with a system that only used uuid-like revisions for atomic writes.  The perf difference I observed in my tests was fairly small though, so it didn't feel worth it.  Option remains open to start doing that in the future if we want though.

**Pull request merged**